### PR TITLE
Create S3 production/cloud processing pipeline and update dagster infrastructure

### LINKED ISSUE #147: Automate local S3 ETL in cloud
### Overview

What is the problem we're solving? For very simple items, this can be encapsulated in the success criteria.

Once #140 is merged, we'll have a local functioning ETL that processes weekly slices of S3 logs. The next step is to get this working in the cloud and on Github actions. For now, we'll stick to using Postgres/CloudSQL, though we'll eventually want to move to BigQuery.

### Success Criteria

How will we know that we're done?
We have a scheduled Github action that runs the latest partition of S3 metrics once a week and updates the CloudSQL database in GCS and in Superset.

```[tasklist]
### Next steps
- [x] Create alembic schema for DB
- [x] Update run_data_update to run last week's partition
- [x] If alembic schema changes - update postgres db
- [x] Get Postgres io manager working for s3 runs
- [x] Get `run_data_update.py` working as expected
- [x] Handle inevitable GCS configuration problems
- [x] Hook up to superset
- [x] Add tempdir / non-temp dir distinction to docs
- [x] Add `partition_key` to `load_input` for SQLiteIOManager and PostgresIOManager
- [x] Run backfill for all S3 data in production
- [x] Handle missing partitions
- [x] Add bogon boolean
- [x] Test GHA to run weekly partition
```

### PR DESCRIPTION
**Pull request by e-belfer**
# Overview

Closes #147.

What problem does this address?

What did you change in this PR?

- Update PostgresIOManager to use dagster's `io_manager` decorator, and create `SQLIOManager` class to hold methods common to both SQL database types
- Add `partition_key` to `load_input` and `handle_output` functions for both IO managers
- Add `parse_dates` handling to prevent dtypes from changing when reading in and out of SQL databases
- Add alembic to handle database schema changes
- Add an environment variable to distinguish between local and prod database runs
- Update `README`
- When writing partitions to SQLite or Postgres databases, only add in records whose primary keys aren't in the dataset yet (this marks partitions as successful rather than failed when they are re-run).
- Update GHA to run `all_metrics_etl` weekly and send update to `pudl-deployments`

# Testing

How did you make sure this worked? How can a reviewer verify this?
Run `all_etl` locally.

# To-do list

```[tasklist]
- [x] Review the PR yourself and call out any questions or issues you have
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,32 +16,32 @@ This is the project structure generated by the [dagster cli](https://docs.dagste
 
 # Setup
 
-## Conda Environment
+## Monda Environment
```
Comment (e-belfer) [inline README.md:19]:
```suggestion
## Mamba Environment
```

File: src/usage_metrics/resources/sqlite.py [hunk 2]
```diff
@@ -49,26 +50,43 @@ def append_df_to_table(self, df: pd.DataFrame, table_name: str) -> None:
...
+                    "All records already loaded, not writing any data. Clobber the database if you want to overwrite this data."
+                )
+            else:
+                df_new.to_sql(
+                    name=table_name,
+                    con=conn,
+                    if_exists="append",
+                    index=False,
+                    dtype={c.name: c.type for c in table_obj.columns},
+                )
```
Comment (bendnorman) [inline src/usage_metrics/resources/sqlite.py:84]: How can we share this logic between the two IO managers so it is only specified in one place?

File: src/usage_metrics/resources/sqlite.py [hunk 3]
```diff
@@ -49,26 +50,43 @@ def append_df_to_table(self, df: pd.DataFrame, table_name: str) -> None:
             table_name in usage_metrics_metadata.tables
         ), f"""{table_name} does not have a database schema defined.
             Create a schema one in usage_metrics.models."""
+        table_obj = usage_metrics_metadata.tables[table_name]
 
         if self.clobber:
```
Comment (bendnorman) [inline src/usage_metrics/resources/sqlite.py:55]: Do we still need the clobber argument now that we are using alembic?

**Review by bendnorman — CHANGES_REQUESTED** [review]
Looking good! I left a few requests and questions for ya.

File: .github/workflows/load-metrics.yml [hunk 4]
```diff
@@ -29,69 +27,52 @@ jobs:
...
           IPINFO_TOKEN: ${{ secrets.IPINFO_TOKEN }}
           POSTGRES_IP: ${{ secrets.POSTGRES_IP }}
           POSTGRES_USER: ${{ secrets.POSTGRES_USER }}
           POSTGRES_PASSWORD: ${{ secrets.POSTGRES_PASSWORD }}
           POSTGRES_DB: ${{ secrets.POSTGRES_DB }}
           POSTGRES_PORT: ${{ secrets.POSTGRES_PORT }}
+        shell: bash -l {0}
         run: |
-          mamba run -n pudl-usage-metrics python run_data_update.py
+          python run_data_update.py
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply .github/workflows/load-metrics.yml:64]: Yes as per the action docs it activates the environment when the shell is initialized: https://github.com/marketplace/actions/setup-micromamba

File: migrations/env.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import logging
+import os
+from logging.config import fileConfig
+
+from alembic import context
+from sqlalchemy import engine_from_config, pool
+
+from usage_metrics.models import usage_metrics_metadata
+from usage_metrics.resources.postgres import PostgresIOManager  # TO DO: Set up paths!!
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply migrations/env.py:11]: Leftover from some earlier experimentation, I'll clear this up!

File: src/usage_metrics/etl/__init__.py [hunk 6]
```diff
@@ -24,6 +27,9 @@
 from usage_metrics.resources.postgres import postgres_manager
 from usage_metrics.resources.sqlite import sqlite_manager
 
+logger = logging.getLogger(__name__)
+warnings.filterwarnings("ignore", category=ExperimentalWarning)
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply src/usage_metrics/etl/__init__.py:31]: `tags` for assets, but this filter isn't working anyways so I'll just drop it.

File: pyproject.toml [hunk 7]
```diff
@@ -9,6 +9,7 @@ dependencies = [
     "sqlalchemy>=2",
     "dagster>=1.7.15, <1.8.3",
     "dagster-webserver>=1.7.15,<1.9",
+    "dagster-duckdb-pandas>=0.24,<0.25",
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply pyproject.toml:12]: Not at all, good catch.

File: migrations/env.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+# ... etc.
+dev_envr = os.getenv("METRICS_PROD_ENV", "local")
+engines = {"prod": PostgresIOManager().engine, "local": SQLiteIOManager().engine}
+
+logger.info(f"Configuring database for {dev_envr} database")
+
+# Get URL from engine based on METRICS_PROD_ENV environment variable and handle escaping %ages
+db_location = (
+    engines[dev_envr].url.render_as_string(hide_password=False).replace("%", "%%")
+)
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply migrations/env.py:45]: Updating env.py to handle this better.

File: src/usage_metrics/resources/postgres.py [hunk 9]
```diff
@@ -51,29 +49,109 @@ def append_df_to_table(self, df: pd.DataFrame, table_name: str) -> None:
...
+            existing_pks = pd.read_sql(sql=query, con=conn)
+            i1 = df.set_index(pk_cols).index
+            i2 = existing_pks.set_index(pk_cols).index
+            # Only update primary keys that aren't in the database
+            df_new = df[~i1.isin(i2)]
+            if df_new.empty:
+                context.log.warn(
+                    "All records already loaded to database. Clobber the database if you want to overwrite this data."
+                )
+            df_new.to_sql(
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply src/usage_metrics/resources/postgres.py:73]: Dagster doesn't distinguish between a successful local and production backfilled partition, so we might end up rerunning partitions by accident sometimes when developing manually. Also could be useful if we run a partition mid-week when testing and want to rerun it automatically at the end of the week without clobbering the whole table.

File: src/usage_metrics/resources/postgres.py [hunk 10]
```diff
@@ -51,29 +49,109 @@ def append_df_to_table(self, df: pd.DataFrame, table_name: str) -> None:
...
+                if context.has_partition_key:
+                    query = query.where(tbl.c["partition_key"] == context.partition_key)
+                df = pd.read_sql(
+                    sql=query,
+                    con=con,
+                    parse_dates=[
+                        col.name
+                        for col in table_obj.columns
+                        if str(col.type) == "TIMESTAMP"
+                    ],
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply src/usage_metrics/resources/postgres.py:133]: The rest of the types seemed to read in and out of Postgres and SQLite correctly when I checked on import/export.

File: src/usage_metrics/resources/sqlite.py (hunk 3 above)
Comment (e-belfer) [reply src/usage_metrics/resources/sqlite.py]: Just went ahead and removed it.

File: src/usage_metrics/resources/sqlite.py (hunk 2 above)
Comment (e-belfer) [reply src/usage_metrics/resources/sqlite.py]: I tried to generalize this using `SQLIOManager` - need to just test that it works on Postgres as expected, but let me know what you think of this strategy?

Comment (e-belfer) [reply src/usage_metrics/resources/sqlite.py]: Tested on the Cloud SQL database and all is working as expected.

File: src/usage_metrics/etl/__init__.py [hunk 11]
```diff
@@ -84,15 +89,17 @@ def _get_keys_from_assets(
...
+    "local": {"database_manager": sqlite_manager},
+}
+
+resources = resources_by_env[os.getenv("METRICS_PROD_ENV", "local")]
 
 defs: Definitions = Definitions(
     assets=default_assets,
     # asset_checks=default_asset_checks,
-    resources={"database_manager": sqlite_manager},  # TODO: How to handle this?
+    resources=resources,  # TODO: How to handle this?
```
Comment (bendnorman) [inline src/usage_metrics/etl/__init__.py:102]: Can these comments be removed?

File: src/usage_metrics/raw/s3.py [hunk 12]
```diff
@@ -73,4 +75,6 @@ def raw_s3_logs(context: AssetExecutionContext) -> pd.DataFrame:
                 weekly_dfs.append(pd.read_csv(path, delimiter=" ", header=None))
             except pd.errors.EmptyDataError:
                 context.log.warnings(f"{path} is an empty file, couldn't read.")
-        return pd.concat(weekly_dfs)
+        if weekly_dfs:  # If data
```
Comment (bendnorman) [inline src/usage_metrics/raw/s3.py:78]: Can you remove this comment or make it more descriptive, please?

File: migrations/env.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bendnorman) [reply migrations/env.py]: This still needs an edit

File: src/usage_metrics/etl/__init__.py [hunk 13]
```diff
@@ -12,7 +13,9 @@
     AssetsDefinition,
     AssetSelection,
     Definitions,
+    ExperimentalWarning,
```
Comment (bendnorman) [inline src/usage_metrics/etl/__init__.py:16]: It doesn't look like this is being used anywhere in the module.

File: src/usage_metrics/resources/sqlite.py (hunk 2 above)
Comment (bendnorman) [reply src/usage_metrics/resources/sqlite.py]: Looks great thank you!

**Review by bendnorman — APPROVED** [review]
Great! Just a few non blocking requests about comments.

**Pull request merged**